lib/optimizer/src/Transformer/ServerSideRendering.php

```diff
@@ -0,0 +1,628 @@
+        ) {
+            return;
+        }
+        $sizer = null;
+
+        if ($layout === Layout::RESPONSIVE) {
+            $sizer = $this->createResponsiveSizer($document, $width, $height);
+        } elseif ($layout === Layout::INTRINSIC) {
+            $sizer = $this->createIntrinsicSizer($document, $width, $height);
+            $this->maybeAddSvgSourceAttributeEncodingFixTweak($document);
+        }
+
+        if ($sizer) {
+            $element->insertBefore($sizer, $element->firstChild);
+        }
+    }
+
+    /**
+     * Create a sizer element for a responsive layout.
+     *
+     * @param Document  $document DOM document to create the sizer for.
+     * @param CssLength $width    Calculated width of the element.
+     * @param CssLength $height   Calculated height of the element.
+     * @return DOMElement
+     */
+    private function createResponsiveSizer(Document $document, CssLength $width, CssLength $height)
+    {
+        $padding = $height->getNumeral() / $width->getNumeral() * 100;
+        $sizer   = $document->createElement(Amp::SIZER_ELEMENT);
+        $sizer->setAttribute(Tag::STYLE, sprintf('display:block;padding-top:%1.4F%%;', $padding));
+
+        return $sizer;
+    }
+
+    /**
+     * Create a sizer element for an intrinsic layout.
+     *
+     * Intrinsic uses an svg inside the sizer element rather than the padding trick.
+     * Note: a naked svg won't work because other things expect the i-amphtml-sizer element.
+     *
+     * @param Document  $document DOM document to create the sizer for.
+     * @param CssLength $width    Calculated width of the element.
+     * @param CssLength $height   Calculated height of the element.
+     * @return DOMElement
+     */
+    private function createIntrinsicSizer(Document $document, CssLength $width, CssLength $height)
+    {
+        $sizer = $document->createElement(Amp::SIZER_ELEMENT);
+        $sizer->setAttribute(Attribute::CLASS_, Amp::SIZER_ELEMENT);
+
+        $sizer_img = $document->createElement(Tag::IMG);
+        $sizer_img->setAttribute(Attribute::ALT, '');
+        $sizer_img->setAttribute(Attribute::ARIA_HIDDEN, 'true');
+        $sizer_img->setAttribute(Attribute::CLASS_, Amp::INTRINSIC_SIZER_ELEMENT);
+        $sizer_img->setAttribute(Attribute::ROLE, 'presentation');
+        $sizer_img->setAttribute(Attribute::SRC, "data:image/svg+xml;charset=utf-8,<svg height=&quot;{$height->getNumeral()}&quot; width=&quot;{$width->getNumeral()}&quot; xmlns=&quot;http://www.w3.org/2000/svg&quot; version=&quot;1.1&quot;/>");
+
+        $sizer->appendChild($sizer_img);
+
+        return $sizer;
+    }
+
+    /**
+     * Maybe add a tweak to the document to fix the SVG source attribute encoding behavior.
+     *
+     * @param Document $document Document instance to add the tweak to.
+     */
+    private function maybeAddSvgSourceAttributeEncodingFixTweak(Document $document)
+    {
+        static $addedSvgSourceAttributeEncodingFixTweak = false;
```

Comment:
I don't think this is right, because the `static` variable is shared across instances. 

```php
class StaticValueContainer {
	function value( $value = null ) {
		static $var;
		if ( $value ) {
			$var = $value;
		}
		return $var;

	}
}

$x = new StaticValueContainer();
$x->value( 1 );

$y = new StaticValueContainer();
$y->value( 2 );

var_dump( $x->value() );
var_dump( $y->value() );
```

This results in:

```
int(2)
int(2)
```

Reply:
Yes, I'm assuming here that there's only one optimizer instantiated per process. It's safer to use a regular property for this.